pilot/pkg/serviceregistry/kube/controller/ambientindex_external.go

```diff
@@ -0,0 +1,376 @@
+			if wl.Waypoint != nil && proto.Equal(wl.Waypoint, addr) {
+				wl.Waypoint = nil
+				// If there was a change, also update the VIPs and record for a push
+				updates.Insert(model.ConfigKey{Kind: kind.Address, Name: wl.ResourceName()})
+			}
+		} else {
+			if wl.Waypoint == nil || !proto.Equal(wl.Waypoint, addr) {
+				wl.Waypoint = addr
+				// If there was a change, also update the VIPs and record for a push
+				updates.Insert(model.ConfigKey{Kind: kind.Address, Name: wl.ResourceName()})
+			}
+		}
+	}
+}
+
+func getWorkloadEntryServices(services []*v1.Service, workloadEntry *apiv1alpha3.WorkloadEntry) []*v1.Service {
+	var filteredServices []*v1.Service
+	for _, service := range services {
+		if service.Spec.Selector == nil {
+			// services with nil selectors match nothing, not everything.
+			continue
+		}
+		if labels.Instance(service.Spec.Selector).SubsetOf(workloadEntry.Labels) {
+			filteredServices = append(filteredServices, service)
+		}
+	}
+	return filteredServices
+}
+
+func findPortForWorkloadEntry(workloadEntry *apiv1alpha3.WorkloadEntry, svcPort *v1.ServicePort) (uint32, error) {
+	if workloadEntry == nil {
+		return 0, fmt.Errorf("invalid input, got nil WorkloadEntry")
+	}
+	if svcPort == nil {
+		return 0, fmt.Errorf("invalid input, got nil ServicePort")
+	}
+
+	for portName, portVal := range workloadEntry.Spec.Ports {
+		if portName == svcPort.Name {
+			return portVal, nil
+		}
+	}
+
+	if svcPort.TargetPort.Type == intstr.Int {
+		return uint32(svcPort.TargetPort.IntValue()), nil
+	}
+
+	return uint32(svcPort.Port), nil
+}
+
+func (c *Controller) getWorkloadEntriesInService(svc *v1.Service) []*apiv1alpha3.WorkloadEntry {
+	allWorkloadEntries := c.getControllerWorkloadEntries(svc.GetNamespace())
+	if svc.Spec.Selector == nil {
+		// services with nil selectors match nothing, not everything.
+		return nil
+	}
+	var workloadEntries []*apiv1alpha3.WorkloadEntry
+	for _, wl := range allWorkloadEntries {
+		if labels.Instance(svc.Spec.Selector).SubsetOf(wl.Spec.Labels) {
+			workloadEntries = append(workloadEntries, wl)
+		}
+	}
+	return workloadEntries
+}
+
+// name format: <cluster>/<group>/<kind>/<namespace>/<name></section-name>
+// if the WorkloadEntry is inlined in the ServiceEntry, we may need section name. caller should use generateServiceEntryUID
+func (c *Controller) generateWorkloadEntryUID(wkEntryNamespace, wkEntryName string) string {
+	return c.clusterID.String() + "/networking.istio.io/WorkloadEntry/" + wkEntryNamespace + "/" + wkEntryName
+}
+
+func (c *Controller) getControllerWorkloadEntries(ns string) []*apiv1alpha3.WorkloadEntry {
+	var allWorkloadEntries []*apiv1alpha3.WorkloadEntry
+	allUnstructuredWorkloadEntries := c.configController.List(gvk.WorkloadEntry, ns)
+	for _, w := range allUnstructuredWorkloadEntries {
+		conv := serviceentry.ConvertWorkloadEntry(w)
+		if conv == nil {
+			continue
+		}
+		c := &apiv1alpha3.WorkloadEntry{
+			ObjectMeta: w.ToObjectMeta(),
+			Spec:       *conv.DeepCopy(),
```

Comment:
Why it is deep copy again, we have shallowcopy in ConvertWorkloadEntry

Reply:
@kdorosh ping

Reply:
Just getting back from PTO. I'll take a look at this and work it into the service entry PR, this was just an abundance of caution but shallow copy should be fine